src/query/ee/src/hilbert_clustering/handler.rs

```diff
@@ -0,0 +1,243 @@
+use databend_common_storages_fuse::SegmentLocation;
+use databend_common_storages_fuse::DEFAULT_BLOCK_PER_SEGMENT;
+use databend_common_storages_fuse::FUSE_OPT_KEY_BLOCK_PER_SEGMENT;
+use databend_enterprise_hilbert_clustering::HilbertClusteringHandler;
+use databend_enterprise_hilbert_clustering::HilbertClusteringHandlerWrapper;
+use databend_storages_common_table_meta::meta::ClusterStatistics;
+use databend_storages_common_table_meta::meta::CompactSegmentInfo;
+use databend_storages_common_table_meta::meta::Statistics;
+use databend_storages_common_table_meta::meta::TableSnapshot;
+
+pub struct RealHilbertClusteringHandler {}
+
+#[async_trait::async_trait]
+impl HilbertClusteringHandler for RealHilbertClusteringHandler {
+    #[async_backtrace::framed]
+    async fn do_hilbert_clustering(
+        &self,
+        table: Arc<dyn Table>,
+        ctx: Arc<dyn TableContext>,
+        push_downs: Option<PushDownInfo>,
+    ) -> Result<Option<(ReclusterInfoSideCar, Arc<TableSnapshot>)>> {
+        let Some((cluster_key_id, _)) = table.cluster_key_meta() else {
+            return Ok(None);
+        };
+
+        let fuse_table = FuseTable::try_from_table(table.as_ref())?;
+        let Some(snapshot) = fuse_table.read_table_snapshot().await? else {
+            // no snapshot, no recluster.
+            return Ok(None);
+        };
+
+        let block_per_seg =
+            fuse_table.get_option(FUSE_OPT_KEY_BLOCK_PER_SEGMENT, DEFAULT_BLOCK_PER_SEGMENT);
+        let block_thresholds = fuse_table.get_block_thresholds();
+        let thresholds = BlockThresholds {
+            max_rows_per_block: block_per_seg * block_thresholds.max_rows_per_block,
+            min_rows_per_block: block_per_seg * block_thresholds.min_rows_per_block,
+            max_bytes_per_block: block_per_seg * block_thresholds.max_bytes_per_block,
+        };
+        let segment_locations = snapshot.segments.clone();
+        let segment_locations = create_segment_location_vector(segment_locations, None);
+
+        let max_threads = ctx.get_settings().get_max_threads()? as usize;
+        let chunk_size = max_threads * 4;
+        let mut checker = ReclusterChecker::new(
+            cluster_key_id,
+            thresholds,
+            push_downs.as_ref().is_none_or(|v| v.filters.is_none()),
+        );
+        'FOR: for chunk in segment_locations.chunks(chunk_size) {
+            // read segments.
+            let compact_segments = FuseTable::segment_pruning(
+                &ctx,
+                fuse_table.schema_with_stream(),
+                fuse_table.get_operator(),
+                &push_downs,
+                fuse_table.get_storage_format(),
+                chunk.to_vec(),
+            )
+            .await?;
+
+            if compact_segments.is_empty() {
+                continue;
+            }
+
+            for (location, segment) in compact_segments.into_iter() {
+                if checker.add(location, segment) {
+                    break 'FOR;
```

Comment:
seems  we do not need label `'FOR`?

Reply:
> seems we do not need label `'FOR`?

It feels like it should be labeled, and if the condition is met, it will just jump out of the outermost loop.

Reply:
u r right, we need break to the label.